docs/http/http-spans.md

```diff
@@ -101,16 +110,32 @@ sections below.
 | `http.response.body.size` | int | The size of the response payload body in bytes. This is the number of bytes transferred excluding headers and is often, but not always, present as the [Content-Length](https://www.rfc-editor.org/rfc/rfc9110.html#field.content-length) header. For requests using transport encoding, this should be the compressed size. | `3495` | Recommended |
-| `http.request.method` | string | HTTP request method. [2] | `GET`; `POST`; `HEAD` | Required |
+| `error.type` | string | Describes a class of error the operation ended with. [2] | `timeout`; `name_resolution_error`; `server_certificate_invalid` | Conditionally Required: If the request has ended with an error. |
+| `http.request.method` | string | HTTP request method. [3] | `GET`; `POST`; `HEAD` | Required |
 | [`network.protocol.name`](../general/attributes.md) | string | [OSI Application Layer](https://osi-model.com/application-layer/) or non-OSI equivalent. The value SHOULD be normalized to lowercase. | `http`; `spdy` | Recommended: if not default (`http`). |
-| [`network.protocol.version`](../general/attributes.md) | string | Version of the application layer protocol used. See note below. [3] | `1.0`; `1.1`; `2`; `3` | Recommended |
-| [`network.transport`](../general/attributes.md) | string | [OSI Transport Layer](https://osi-model.com/transport-layer/) or [Inter-process Communication method](https://en.wikipedia.org/wiki/Inter-process_communication). The value SHOULD be normalized to lowercase. | `tcp`; `udp` | Conditionally Required: [4] |
+| [`network.protocol.version`](../general/attributes.md) | string | Version of the application layer protocol used. See note below. [4] | `1.0`; `1.1`; `2`; `3` | Recommended |
+| [`network.transport`](../general/attributes.md) | string | [OSI Transport Layer](https://osi-model.com/transport-layer/) or [Inter-process Communication method](https://en.wikipedia.org/wiki/Inter-process_communication). The value SHOULD be normalized to lowercase. | `tcp`; `udp` | Conditionally Required: [5] |
 | [`network.type`](../general/attributes.md) | string | [OSI Network Layer](https://osi-model.com/network-layer/) or non-OSI equivalent. The value SHOULD be normalized to lowercase. | `ipv4`; `ipv6` | Recommended |
 | `user_agent.original` | string | Value of the [HTTP User-Agent](https://www.rfc-editor.org/rfc/rfc9110.html#field.user-agent) header sent by the client. | `CERN-LineMode/2.15 libwww/2.17b3` | Recommended |
 
 **[1]:** If and only if it's different than `http.request.method`.
 
-**[2]:** HTTP request method value SHOULD be "known" to the instrumentation.
+**[2]:** If the response status code was sent/received and the status indicates an error according to the [HTTP span status definition](/docs/http/http-spans.md),
```

Comment:
Does this also apply to the client metrics? If so, I wonder why it has been decided to report the error redundantly, when it can be implied from the status code? In case of the client, it makes more sense to me to report local errors in `error.type` and remote errors in `status_code` like @noahfalk recommended in https://github.com/open-telemetry/semantic-conventions/pull/205#issuecomment-1649153582.

> and the status indicates an error according to the [HTTP span status definition](https://github.com/open-telemetry/semantic-conventions/blob/a6a7593932dcd1d13e07f296fd429ae36fa688bf/docs/http/http-spans.md)

It's also weird that this document is self-referencing without pointing to a particular section.

Reply:
> the reasoning we had in mind is so you can report on http errors via a single attribute, instead of having to look at two different attributes

The downside is that it's mixing concerns. These are two very distinct classes of errors, at least on the client. Separating them makes it easier to exclude one or the other class when analyzing metrics.

Reply:
@antonfirsov @noahfalk Is reason phrase the contentious part here?
- would you like to record something else when the status code is reported?  (BTW, nothing will overwrite `error.type` set by the instrumentation)
- we can make a recommendation to use reason phrase softer (but still, it MUST NOT be empty to count as an error)

Redundancy is not really a problem here since it does not introduce new time series.

As @trask mentioned the goal is to give users one attribute to measure error rate with query like `rate(http_request_duration_count{error_type!=""}[1m]) by error_type`. Users can still group by status code. Error types still heavily depend on the library.

Multiple attributes would lead to a worse user experience: to get error rate I need to look at multiple attributes and write query like 
`rate(http_request_duration_count{error_type!="",http_response_status_code=~"200|201|302"}[1m]) by error_type, http_response_status_code`

There is nothing that users are losing if a reason phrase (or at least `_OTHER`) is reported in the `error.type`, but they are gaining a much better query experience.

Reply:
and if users want to only see errors without response they can still write `rate(http_request_duration_count{error_type!="",http_response_status_code=""}[1m]) by error_type` which is much easier than trying to filter out 'successful' response codes

Reply:
> There is nothing that users lose if a reason phrase (or at least _OTHER) is reported in the error.type

I think they could lose something if I understood the proposal properly? Imagine someone writes a library where every time a request comes back with status code 404, the library throws exception RequestNotFoundException. The library also throws DnsException when DNS lookup fails, and InvalidResponseException when the message is malformed. So engineers go write some queries error_type=="RequestNotFoundException", error_type=="DnsException", and error_type=="InvalidResponseException". Did I understand the proposal properly that the last two queries work fine, but the first one doesn't work because those errors report the reason phrase rather than the library's representation of the error?

That said, I think I am coming around that at least for .NET the ambiguity introduced by the proposal would be minimal. .NET doesn't throw exceptions based on response status code like that example above (please correct me if I am mistaken @antonfirsov). This means union of both error types in the same attribute appears non-destructive at least for us. There is a narrow scenario where a server-side error occurs and a different client-side error occurs afterwards, but I am guessing that is rare is enough that it wouldn't matter if a query mis-triages it. Regardless whether errors are merged in the same attribute or not, I think the query for server-side errors looks like `http_request_duration_count{http_response_status_code=~"4.+|5.+"}`. The queries that change are the query for all-errors and the query for client-side-only errors. If the errors are merged the all-errors query is simple and client-only is more complicated. If the errors aren't merged the situation swaps, client-only is simple, all-errors is more complicated. I think its fair that the all-errors query is more commonly desired than the client-only errors query so that one gets prioritized as the simple one.

@antonfirsov - you may still have concerns about usability and glad to continue discussing, but I just wanted to check if you have any concerns if this would be difficult to implement as currently proposed? I'm guessing that aspect is OK?

Reply:
Note that the reason phrase used to be contained in a very old version of OTel HTTP semantic conventions but was removed because it was deemed useless (and I would agree to that, generally).
https://github.com/open-telemetry/opentelemetry-specification/pull/972

Reply:
> .NET doesn't throw exceptions based on response status code

HttpClient methods which are expected to process the response body (`GetString` etc.) do throw, but we don't log those exceptions in `MetricsHandler` since `MetricsHandler` does not know which `HttpClient` method was invoked. It's job finishes after *headers read*.

> any concerns if this would be difficult to implement

Ideally we should log `error.type` whenever an exception happens, but this is not possible for `GetString` etc. methods mentioned above. Which opens the question when should we log it? In all cases where status code is not 2**, even though `HttpClient` didn't throw?

> @antonfirsov - you may still have concerns about usability

Your analysis looks correct and looks like the transformation is non-destructive, therefore it's likely possible to overcome usability problems. My concern is that from implementation perspective the definition of `error.type` feels unpractical and forces us to make arbitrary decisions (see previous note). Maybe I don't understand the consumer side well enough, but it feels like having a single attribute is only beneficial for beginners, and a more advanced user may prefer a stricter separation between the error classes without the need of writing custom queries.

Reply:
@antonfirsov 

> HttpClient methods which are expected to process the response body (GetString etc.) do throw, but we don't log those exceptions in MetricsHandler since MetricsHandler does not know which HttpClient method was invoked. Its' job finishes after headers read.

Does it mean that on .NET client span duration/outcome and what's recorded on metrics are not the same? It's a great discussion, but goes beyond this PR

> Ideally we should log error.type whenever an exception happens, but this is not possible for GetString etc. methods mentioned above. Which opens the question when should we log it? In all cases where status code is not 2**, even though HttpClient didn't throw?

Based on the current proposal here (see examples for spans), `200` and `SocketException` can coexist and seem quite useful. So we share the same *ideal* vision. Do you believe .NET will never be able to achieve it because of hard limitations of some sort? 

>  My concern is that from implementation perspective the definition of error.type feels unpractical and forces us to make arbitrary decisions (see previous note). Maybe I don't understand the consumer side well enough, but it feels like having a single attribute is only beneficial for beginners, and a more advanced user may prefer a stricter separation between the error classes without the need of writing custom queries.

What would be the limitations for advanced users? If they need to query by `error.type` when no response is recieved/sent, they can still use `rate(http_request_duration_count{error_type!="",http_response_status_code=""}[1m]) by error_type` and write any advanced queries that filter out certain  `http_response_status_code`.
But then we only ask advanced users to write advanced queries, with two attributes we're asking everyone to write advanced queries.

Reply:
> Such a library has strong reasons to report RequestNotFoundException in the error.type. Reason-phrase is under SHOULD clause and strong reasons are stronger than it

It sounds like I just misunderstood that aspect of the proposal, my bad. No need to change the wording of the proposal on my account, but if you think it will help others understand better I'm happy to leave that wording choice up to you.

> Which opens the question when should we log it? In all cases where status code is not 2**, even though HttpClient didn't throw?

I think the proposal works out to:
- If there is an Exception in flight through the MetricsHandler, keep the behavior that is currently implemented (error.type is Exception name or error.type is HttpRequestError enumeration)
- If there is no Exception in flight at that point and response code is 4xx or 5xx, use the status code as the error
- Otherwise error.type is empty

> My concern is that from implementation perspective the definition of error.type feels unpractical and forces us to make arbitrary decisions (see previous note)

Does that sketch of the implementation above feel impractical? The decision didn't feel arbitrary to me. It seems derivable by prioritizing which user queries would be most common, then chose a telemetry shape that optimizes making the most common query simple. Is the part you found arbitrary the definition of which conditions users are most likely to consider as errors or something else?

> a more advanced user may prefer a stricter separation between the error classes without the need of writing custom queries

I'm not sure if you mean 'custom' as in the dev wrote it themselves, or 'custom' as in 'a query that exceeds a certain threshold of complexity'? If you meant don't require the dev to write the query, we would achieve that goal by having someone (us, APM vendors, OSS contributor) create a dashboard in advance and then everyone else uses it. That is always an option regardless of which proposal we go with. I'm guessing you meant 'custom' as in queries that are more complex though.

If we have three queries we are interested in (all-errors, server-side-errors, client-side-errors), any design that encodes the error info in only two attributes requires at least one of the those queries to reference both attributes and thus be more complex. If we want to make the client-errors and server-errors queries both be simple, the tradeoff is that the more common all-errors query, used by both beginners and advanced users, is the complicated one. That doesn't seem like a good tradeoff. It is possible to make all the queries simple by using three attributes (client-side-error, server-side-error, merged-error), and then I think its a judgement call of does the extra simplicity gained for client-side-error query justify having a 3rd attribute dedicated to only client-side errors? I assume the OTel folks believe that would not be a good tradeoff and at this point I think I agree. I was initially worried the merger was designed to be lossy, but that was just my misunderstanding.

Reply:
> I'm guessing you meant 'custom' as in queries that are more complex though.

Yes, that was bad wording on my side.

> If the response status code was sent/received and the status indicates an error according to the [HTTP span status definition](https://github.com/docs/http/http-spans.md#status),
error.type SHOULD be set to the string representation of the status code, type of exception thrown by the component (if any), or a component-specific low-cardinality error code. If none of these options apply, it MUST be set to _OTHER.

Ok, I'm actually fine with this definition given @noahfalk 's arguments in https://github.com/open-telemetry/semantic-conventions/pull/205#discussion_r1319149110. 👍

> Does it mean that on .NET client span duration/outcome and what's recorded on metrics are not the same? It's a great discussion, but goes beyond this PR

From metrics perspective the _HTTP client span_ lasts until "headers read", regardless of the top-level `HttpClient` call which may or may not buffer the response content afterwards. This has been decided after long discussions on various forums, and my understanding was that this conforms the [HTTP client span duration](https://github.com/open-telemetry/semantic-conventions/blob/03db37ed7c06c05ebabb28e4abec46745a96e92d/docs/http/http-spans.md#http-client-span-duration) spec. If this was a misunderstanding, we should open a discussion somewhere.

Reply:
> This has been decided after long discussions on various forums, and my understanding was that this conforms the [HTTP client span duration](https://github.com/open-telemetry/semantic-conventions/blob/03db37ed7c06c05ebabb28e4abec46745a96e92d/docs/http/http-spans.md#http-client-span-duration) spec we ended up with. If this was a misunderstanding, we should open a discussion somewhere.

@antonfirsov, Thanks for the clarification!

Measuring duration this way is fine (I need to make sure it's documented in #283), as long as span duration and metric measurement are consistent

https://github.com/open-telemetry/semantic-conventions/blob/ff200826aec224851e754cefe1c857d4f47e36a5/docs/http/http-metrics.md?plain=1#L369-L375

It all comes with SHOULD and I don't see not following it as a showstopper, but eventually, it'd be nice to align span duration/status and measurement duration/error.type. If you believe it can never be done, let's discuss - it's a good input for OTel HTTP semconv.

> I think its a judgement call of does the extra simplicity gained for client-side-error query justify having a 3rd attribute dedicated to only client-side errors? I assume the OTel folks believe that would not be a good tradeoff and at this point I think I agree. I was initially worried the merger was designed to be lossy, but that was just my misunderstanding.

The beauty of a 3rd attribute is that we should be able to add it later in a non-breaking manner based on the feedback.

Reply:
[from now on this is off topic]

> it'd be nice to align span duration/status and measurement duration/error.type

@lmolkova what/who defines the start and the finish of an http client span? My thinking was that [http-spans.md#http-client-span-duration](https://github.com/open-telemetry/semantic-conventions/blob/03db37ed7c06c05ebabb28e4abec46745a96e92d/docs/http/http-spans.md#http-client-span-duration) leaves it to the library docs to clarify, and it doesn't have to align with the span of the top-level API call. In my  interpretation .NET HttpClient's span ends at headers finished regardless of `HttpCompletionOption`, which aligns with the duration measurement.

Reply:
I updated the PR clarifying exceptions/lib-specific errors when the response is received. PTAL